[ISSUE] Issue with `databricks_schema` resource - "UpdateSchema Nothing to update"

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Databricks Provider issue tracker reserved for bug reports and feature requests. For general usage questions, please see: <https://www.terraform.io/community.html>.
-->
Terraform returns an error when updating `force_destroy` or removing `comment` on the `databricks_schema` resource. Based on the debug logs, It looks like the issue occurs because the provider passes an empty object to the PATCH endpoint for schemas.

### Configuration
<!-- Please provide a minimal reproducible configuration for the issue -->
```hcl
resource "databricks_catalog" "this" {
  name = "force_destroy_test_catalog"
}

resource "databricks_schema" "this" {
  name          = "force_destroy_test_schema"
  catalog_name  = databricks_catalog.this.name
  comment       = "Test"
  force_destroy = true
}
```

### Expected Behavior
<!-- What should have happened? -->
- Terraform should apply changes to `force_destroy` and removal of `comment` without errors.
- The schema comment should be deleted (or set to an empty string) on the actual resource.

### Actual Behavior
Terraform returns an error message:

> Error: cannot update schema: UpdateSchema Nothing to update.

- `force_destroy` is properly updated in Terraform state
- `comment` is properly updated in Terraform state but *not* in the actual resource. Subsequent `terraform plan` commands do not detect the difference between TF state and reality. 

### Steps to Reproduce
<!-- Please list the steps required to reproduce the issue, for example:-->
1. `terraform apply`
2. Comment out or remove `comment` and `force_destroy`
3. `terraform apply`

### Terraform and provider versions
<!-- Please paste the output of `terraform version`. If version of `databricks` provider is not the latest (https://github.com/databricks/terraform-provider-databricks/releases), please make sure to use the latest one. -->
Terraform v1.7.5
on darwin_amd64
+ provider registry.terraform.io/databricks/databricks v1.52.0
+ provider registry.terraform.io/hashicorp/azurerm v3.116.0

### Is it a regression?
<!-- Did this work in a previous version of the provider? If so, which versions did you try? -->
Not sure.

### Debug Output
<!-- Please add turn on logging, e.g. `TF_LOG=DEBUG terraform apply -no-color` and run command again, paste it to gist & provide the link to gist. If you're still willing to paste in log output, make sure you provide only relevant log lines with requests. It would make it more readable, if you pipe the log through `| grep databricks | sed -E 's/^.* plugin[^:]+: (.*)$/\1/'`, e.g.: `TF_LOG=DEBUG terraform apply -no-color 2>&1 | grep databricks | sed -E 's/^.* plugin[^:]+: (.*)$/\1/' 2>&1 |tee tf-debug.log`. If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. -->
```
2024-09-25T18:04:28.335+0200 [DEBUG] provider.terraform-provider-databricks_v1.52.0: PATCH /api/2.1/unity-catalog/schemas/force_destroy_test_catalog.force_destroy_test_schema
> {}
< HTTP/2.0 400 Bad Request
< {
<   "details": [
<     {
<       "@type": "type.googleapis.com/google.rpc.ErrorInfo",
<       "domain": "unity-catalog.databricks.com",
<       "metadata": {
<         "field_name": "schema_info"
<       },
<       "reason": "INVALID_FIELD"
<     },
<     {
<       "@type": "type.googleapis.com/google.rpc.RequestInfo",
<       "request_id": "2545238d-6d1c-4472-83a7-3233c9b0af12",
<       "serving_data": ""
<     }
<   ],
<   "error_code": "INVALID_PARAMETER_VALUE",
<   "message": "UpdateSchema Nothing to update."
< }: tf_req_id=bdf3b3df-83b5-724b-3e65-1cba60664517 @caller=/home/runner/work/terraform-provider-databricks/terraform-provider-databricks/logger/logger.go:38 tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/databricks/databricks @module=databricks tf_rpc=ConfigureProvider timestamp="2024-09-25T18:04:28.335+0200"
2024-09-25T18:04:28.336+0200 [DEBUG] provider.terraform-provider-databricks_v1.52.0: non-retriable error: UpdateSchema Nothing to update.: tf_mux_provider=tf5to6server.v5tov6Server tf_provider_addr=registry.terraform.io/databricks/databricks tf_req_id=bdf3b3df-83b5-724b-3e65-1cba60664517 @caller=/home/runner/work/terraform-provider-databricks/terraform-provider-databricks/logger/logger.go:38 @module=databricks tf_rpc=ConfigureProvider timestamp="2024-09-25T18:04:28.336+0200"
2024-09-25T18:04:28.336+0200 [ERROR] provider.terraform-provider-databricks_v1.52.0: Response contains error diagnostic: @caller=/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/internal/diag/diagnostics.go:58 tf_req_id=ae776571-ea6c-43e4-674e-97d782b905ec @module=sdk.proto tf_resource_type=databricks_schema diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="cannot update schema: UpdateSchema Nothing to update." tf_rpc=ApplyResourceChange tf_proto_version=6.6 tf_provider_addr=registry.terraform.io/databricks/databricks timestamp="2024-09-25T18:04:28.336+0200"
```

### Important Factoids
<!-- Are there anything atypical about your accounts that we should know? -->
- Using Azure Databricks

### Would you like to implement a fix?
<!-- If you plan to implement a fix for this, let the maintainers and community know -->
No

--- a/catalog/resource_catalog.go
+++ b/catalog/resource_catalog.go
@@ -124,6 +124,9 @@ func ResourceCatalog() common.Resource {
 			if err != nil {
 				return err
 			}
+			if !d.HasChangeExcept("force_destroy") {
+				return nil
+			}
 
 			var updateCatalogRequest catalog.UpdateCatalog
 			common.DataToStructPointer(d, catalogSchema, &updateCatalogRequest)
@@ -141,6 +144,10 @@ func ResourceCatalog() common.Resource {
 
 			if !d.HasChangeExcept("owner") {
 				return nil
+			}
+
+			if d.HasChange("comment") && updateCatalogRequest.Comment == "" {
+				updateCatalogRequest.ForceSendFields = append(updateCatalogRequest.ForceSendFields, "Comment")
 			}
 
 			updateCatalogRequest.Owner = ""

--- a/catalog/resource_catalog_test.go
+++ b/catalog/resource_catalog_test.go
@@ -227,6 +227,67 @@ func TestUpdateCatalog(t *testing.T) {
 	}.ApplyNoError(t)
 }
 
+func TestUpdateCatalogSetEmptyComment(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {
+			w.GetMockMetastoresAPI().EXPECT().Current(mock.Anything).Return(&catalog.MetastoreAssignment{
+				MetastoreId: "d",
+			}, nil)
+			e := w.GetMockCatalogsAPI().EXPECT()
+			e.Update(mock.Anything, catalog.UpdateCatalog{
+				Name:            "a",
+				Comment:         "",
+				ForceSendFields: []string{"Comment"},
+			}).Return(&catalog.CatalogInfo{
+				Name:    "a",
+				Comment: "",
+			}, nil)
+			e.GetByName(mock.Anything, "a").Return(&catalog.CatalogInfo{
+				Name:    "a",
+				Comment: "",
+			}, nil)
+		},
+		Resource: ResourceCatalog(),
+		Update:   true,
+		ID:       "a",
+		InstanceState: map[string]string{
+			"metastore_id": "d",
+			"name":         "a",
+			"comment":      "c",
+		},
+		HCL: `
+		name = "a"
+		comment = ""
+		`,
+	}.ApplyNoError(t)
+}
+
+func TestUpdateCatalogForceDestroyOnly(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {
+			w.GetMockMetastoresAPI().EXPECT().Current(mock.Anything).Return(&catalog.MetastoreAssignment{
+				MetastoreId: "d",
+			}, nil)
+			e := w.GetMockCatalogsAPI().EXPECT()
+			e.GetByName(mock.Anything, "a").Return(&catalog.CatalogInfo{
+				Name: "a",
+			}, nil)
+		},
+		Resource: ResourceCatalog(),
+		Update:   true,
+		ID:       "a",
+		InstanceState: map[string]string{
+			"metastore_id":  "d",
+			"name":          "a",
+			"force_destroy": "true",
+		},
+		HCL: `
+		name = "a"
+		force_destroy = false
+		`,
+	}.ApplyNoError(t)
+}
+
 func TestUpdateCatalogOwnerOnly(t *testing.T) {
 	qa.ResourceFixture{
 		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {

--- a/catalog/resource_registered_model.go
+++ b/catalog/resource_registered_model.go
@@ -92,6 +92,9 @@ func ResourceRegisteredModel() common.Resource {
 				return nil
 			}
 
+			if d.HasChange("comment") && u.Comment == "" {
+				u.ForceSendFields = append(u.ForceSendFields, "Comment")
+			}
 			u.Owner = ""
 			_, err = w.RegisteredModels.Update(ctx, u)
 			if err != nil {

--- a/catalog/resource_registered_model_test.go
+++ b/catalog/resource_registered_model_test.go
@@ -196,6 +196,47 @@ func TestRegisteredModelUpdate(t *testing.T) {
 	}.ApplyNoError(t)
 }
 
+func TestRegisteredModelUpdateCommentOnly(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {
+			e := w.GetMockRegisteredModelsAPI().EXPECT()
+			e.Update(mock.Anything, catalog.UpdateRegisteredModelRequest{
+				FullName:        "catalog.schema.model",
+				Comment:         "",
+				ForceSendFields: []string{"Comment"},
+			}).Return(&catalog.RegisteredModelInfo{
+				Name:        "model",
+				CatalogName: "catalog",
+				SchemaName:  "schema",
+				FullName:    "catalog.schema.model",
+				Comment:     "",
+			}, nil)
+			e.GetByFullName(mock.Anything, "catalog.schema.model").Return(&catalog.RegisteredModelInfo{
+				Name:        "model",
+				CatalogName: "catalog",
+				SchemaName:  "schema",
+				FullName:    "catalog.schema.model",
+				Comment:     "",
+			}, nil)
+		},
+		Resource: ResourceRegisteredModel(),
+		Update:   true,
+		ID:       "catalog.schema.model",
+		InstanceState: map[string]string{
+			"name":         "model",
+			"catalog_name": "catalog",
+			"schema_name":  "schema",
+			"comment":      "comment",
+		},
+		HCL: `
+			name = "model"
+			catalog_name = "catalog"
+			schema_name = "schema"
+			comment = ""
+			`,
+	}.ApplyNoError(t)
+}
+
 func TestRegisteredModelUpdateOwner(t *testing.T) {
 	qa.ResourceFixture{
 		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {

--- a/catalog/resource_schema.go
+++ b/catalog/resource_schema.go
@@ -100,6 +100,9 @@ func ResourceSchema() common.Resource {
 			if err != nil {
 				return err
 			}
+			if !d.HasChangeExcept("force_destroy") {
+				return nil
+			}
 			var updateSchemaRequest catalog.UpdateSchema
 			common.DataToStructPointer(d, s, &updateSchemaRequest)
 			updateSchemaRequest.FullName = d.Id()
@@ -116,6 +119,10 @@ func ResourceSchema() common.Resource {
 
 			if !d.HasChangeExcept("owner") {
 				return nil
+			}
+
+			if d.HasChange("comment") && updateSchemaRequest.Comment == "" {
+				updateSchemaRequest.ForceSendFields = append(updateSchemaRequest.ForceSendFields, "Comment")
 			}
 
 			updateSchemaRequest.Owner = ""

--- a/catalog/resource_schema_test.go
+++ b/catalog/resource_schema_test.go
@@ -135,6 +135,77 @@ func TestUpdateSchema(t *testing.T) {
 	}.ApplyNoError(t)
 }
 
+func TestUpdateSchemaSetEmptyComment(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {
+			w.GetMockMetastoresAPI().EXPECT().Current(mock.Anything).Return(&catalog.MetastoreAssignment{
+				MetastoreId: "d",
+			}, nil)
+			e := w.GetMockSchemasAPI().EXPECT()
+			e.Update(mock.Anything, catalog.UpdateSchema{
+				FullName:        "b.a",
+				Comment:         "",
+				ForceSendFields: []string{"Comment"},
+			}).Return(&catalog.SchemaInfo{
+				FullName: "b.a",
+				Owner:    "administrators",
+			}, nil)
+			e.GetByFullName(mock.Anything, "b.a").Return(&catalog.SchemaInfo{
+				Name:        "a",
+				CatalogName: "b",
+				MetastoreId: "d",
+				Owner:       "administrators",
+			}, nil)
+		},
+		Resource: ResourceSchema(),
+		Update:   true,
+		ID:       "b.a",
+		InstanceState: map[string]string{
+			"metastore_id": "d",
+			"name":         "a",
+			"catalog_name": "b",
+			"comment":      "c",
+		},
+		HCL: `
+		name = "a"
+		catalog_name = "b"
+		`,
+	}.ApplyAndExpectData(t, map[string]any{
+		"comment": "",
+	})
+}
+
+func TestUpdateSchemaChangeForceDestroy(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {
+			w.GetMockMetastoresAPI().EXPECT().Current(mock.Anything).Return(&catalog.MetastoreAssignment{
+				MetastoreId: "d",
+			}, nil)
+			e := w.GetMockSchemasAPI().EXPECT()
+			e.GetByFullName(mock.Anything, "b.a").Return(&catalog.SchemaInfo{
+				Name:        "a",
+				CatalogName: "b",
+				MetastoreId: "d",
+				Owner:       "administrators",
+			}, nil)
+		},
+		Resource: ResourceSchema(),
+		Update:   true,
+		ID:       "b.a",
+		InstanceState: map[string]string{
+			"metastore_id":  "d",
+			"name":          "a",
+			"catalog_name":  "b",
+			"force_destroy": "true",
+		},
+		HCL: `
+		name = "a"
+		catalog_name = "b"
+		force_destroy = false
+		`,
+	}.ApplyNoError(t)
+}
+
 func TestUpdateSchemaOwnerWithOtherFields(t *testing.T) {
 	qa.ResourceFixture{
 		MockWorkspaceClientFunc: func(w *mocks.MockWorkspaceClient) {

--- a/catalog/resource_volume.go
+++ b/catalog/resource_volume.go
@@ -127,6 +127,10 @@ func ResourceVolume() common.Resource {
 				return nil
 			}
 
+			if d.HasChange("comment") && updateVolumeRequestContent.Comment == "" {
+				updateVolumeRequestContent.ForceSendFields = append(updateVolumeRequestContent.ForceSendFields, "Comment")
+			}
+
 			updateVolumeRequestContent.Owner = ""
 			v, err := w.Volumes.Update(ctx, updateVolumeRequestContent)
 			if err != nil {

--- a/catalog/resource_volume_test.go
+++ b/catalog/resource_volume_test.go
@@ -359,6 +359,65 @@ func TestVolumesUpdate(t *testing.T) {
 	assert.Equal(t, "/Volumes/testCatalogName/testSchemaName/testNameNew", d.Get("volume_path"))
 }
 
+func TestVolumesUpdateCommentOnly(t *testing.T) {
+	qa.ResourceFixture{
+		Fixtures: []qa.HTTPFixture{
+			{
+				Method:   http.MethodPatch,
+				Resource: "/api/2.1/unity-catalog/volumes/testCatalogName.testSchemaName.testName",
+				ExpectedRequest: catalog.UpdateVolumeRequestContent{
+					Comment:         "",
+					ForceSendFields: []string{"Comment"},
+				},
+				Response: catalog.VolumeInfo{
+					Name:        "testName",
+					VolumeType:  catalog.VolumeType("testVolumeType"),
+					CatalogName: "testCatalogName",
+					SchemaName:  "testSchemaName",
+					Comment:     "",
+					FullName:    "testCatalogName.testSchemaName.testName",
+				},
+			},
+			{
+				Method:   http.MethodGet,
+				Resource: "/api/2.1/unity-catalog/volumes/testCatalogName.testSchemaName.testName?",
+				Response: catalog.VolumeInfo{
+					Name:        "testName",
+					VolumeType:  catalog.VolumeType("testVolumeType"),
+					CatalogName: "testCatalogName",
+					SchemaName:  "testSchemaName",
+					Comment:     "",
+					FullName:    "testCatalogName.testSchemaName.testNameNew",
+				},
+			},
+		},
+		Resource: ResourceVolume(),
+		Update:   true,
+		InstanceState: map[string]string{
+			"name":         "testName",
+			"catalog_name": "testCatalogName",
+			"schema_name":  "testSchemaName",
+			"volume_type":  "testVolumeType",
+			"comment":      "this is a comment",
+		},
+		ID: "testCatalogName.testSchemaName.testName",
+		HCL: `
+		name = "testName"
+		volume_type = "testVolumeType"
+		catalog_name = "testCatalogName"
+		schema_name = "testSchemaName"
+		comment = ""
+		`,
+	}.ApplyAndExpectData(t, map[string]any{
+		"name":         "testName",
+		"volume_type":  "testVolumeType",
+		"catalog_name": "testCatalogName",
+		"schema_name":  "testSchemaName",
+		"comment":      "",
+		"volume_path":  "/Volumes/testCatalogName/testSchemaName/testNameNew",
+	})
+}
+
 func TestVolumesUpdateForceNewOnCatalog(t *testing.T) {
 	d, err := qa.ResourceFixture{
 		Fixtures: []qa.HTTPFixture{

